src/diagram.rs

```diff
@@ -0,0 +1,807 @@
+    /// }
+    /// "#;
+    ///
+    /// let diagram = Diagram::from_json_str(json_str)?;
+    /// let workflow = diagram.spawn_io_workflow(&mut app, &registry)?;
+    /// # Ok::<_, DiagramError>(())
+    /// ```
+    // TODO(koonpeng): Support streams other than `()` #43.
+    /* pub */
+    fn spawn_workflow<Streams>(
+        &self,
+        cmds: &mut Commands,
+        registry: &NodeRegistry,
+    ) -> Result<Service<DiagramStart, DiagramTerminate, Streams>, DiagramError>
+    where
+        Streams: StreamPack,
+    {
+        let mut err: Option<DiagramError> = None;
+
+        macro_rules! unwrap_or_return {
+            ($v:expr) => {
+                match $v {
+                    Ok(v) => v,
+                    Err(e) => {
+                        err = Some(e);
+                        return;
+                    }
+                }
+            };
+        }
+
+        let w = cmds.spawn_workflow(|scope: DiagramScope<Streams>, builder: &mut Builder| {
+            debug!(
+                "spawn workflow, scope input: {:?}, terminate: {:?}",
+                scope.input.id(),
+                scope.terminate.id()
+            );
+
+            unwrap_or_return!(create_workflow(scope, builder, registry, self));
+        });
+
+        if let Some(err) = err {
+            return Err(err);
+        }
+
+        Ok(w)
+    }
+
+    /// Wrapper to [spawn_workflow::<()>](Self::spawn_workflow).
+    pub fn spawn_io_workflow(
+        &self,
+        cmds: &mut Commands,
+        registry: &NodeRegistry,
+    ) -> Result<Service<DiagramStart, DiagramTerminate, ()>, DiagramError> {
+        self.spawn_workflow::<()>(cmds, registry)
+    }
+
+    pub fn from_json(value: serde_json::Value) -> Result<Self, serde_json::Error> {
+        serde_json::from_value(value)
+    }
+
+    pub fn from_json_str(s: &str) -> Result<Self, serde_json::Error> {
+        serde_json::from_str(s)
+    }
+
+    pub fn from_reader<R>(r: R) -> Result<Self, serde_json::Error>
+    where
+        R: Read,
+    {
+        serde_json::from_reader(r)
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DiagramError {
+    #[error("node builder [{0}] is not registered")]
+    BuilderNotFound(BuilderId),
+
+    #[error("operation [{0}] not found")]
+    OperationNotFound(OperationId),
+
+    #[error("output type does not match input type")]
+    TypeMismatch,
+
+    #[error("missing start or terminate")]
+    MissingStartOrTerminate,
+
+    #[error("cannot connect to start")]
+    CannotConnectStart,
+
+    #[error("request or response cannot be serialized or deserialized")]
+    NotSerializable,
+
+    #[error("response cannot be cloned")]
+    NotCloneable,
+
+    #[error("the number of unzip slots in response does not match the number of inputs")]
+    NotUnzippable,
+
+    #[error(
+        "node must be registered with \"with_fork_result()\" to be able to perform fork result"
+    )]
+    CannotForkResult,
+
+    #[error("response cannot be splitted")]
```

Comment:
```suggestion
    #[error("response cannot be split")]
```

split is one of those weird English words where it is its own past tense

Reply:
bf63e5b20290473dbc34dc95a1e0057422d8499d